File: 백준/Gold/1987. 알파벳/알파벳.py
r, c = map(int, input().split()); cnt = 0; chars = 0; length = 0
arr = [input() for _ in range(r)]; directions = [[0,-1],[0,1],[-1,0],[1,0]]
def go(y, x):
	global cnt, chars, length
	if cnt > ord('z') - ord('a'): return # 모든 알파벳 돎
	if y < 0 or x < 0 or y >= r or x >= c: return # 범위 벗어남
	if chars & (1 << (ord(arr[y][x]) - ord('A'))): return # 이미 거친 알파벳일 때
	chars |= (1 << ord(arr[y][x]) - ord('A'))
	length += 1; cnt = max(cnt, length)
	for d in directions:
		dy = y + d[0]; dx = x + d[1]
		go(dy, dx)
	chars -= (1 << ord(arr[y][x]) - ord('A')); length -= 1
go(0, 0)
print(cnt)

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.util.*;

public class Main {
	public static int[] rc;
	public static char[][] arr;
	public static int cnt = 0, chars = 0, length = 0;
	public static int[][] directions = {{0,1},{1,0},{-1,0},{0,-1}};
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		rc = Arrays.stream(scanner.nextLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		arr = new char[rc[0]][rc[1]];
		for (int i = 0; i < rc[0]; i++) arr[i] = scanner.nextLine().toCharArray();
		go(0, 0);
		System.out.print(cnt);
	}

	public static void go(int y, int x) {
		if(y < 0 || x < 0 || y >= rc[0] || x >= rc[1]) return;
		if(cnt > 'z' - 'a') return;
		int toBinary = 1 << arr[y][x];
        if((chars & toBinary) > 0) return;
		length++; chars |= toBinary; cnt = Math.max(cnt, length);
		for (int[] d : directions) {
			int dy = y + d[0], dx = x + d[1]; go(dy, dx);
		}
		length--; chars -= toBinary;
	}
}